### ISSUE TITLE
Custom user agent functionality?

### ISSUE THREAD

**Issue by chrismash2** [openid/AppAuth-Android#544]
Hi,

I've been using the iOS version of this library for a little while and making use of the [custom user agent](https://github.com/openid/AppAuth-iOS#custom-user-agents) functionality to handle the authentication, though I can't see the same functionality available on Android, has it never been implemented? Is there a way to achieve the same behaviour without modifying the library?

**Comment by iainmcgin** [comment]
The user agent can be customized for token requests by using a custom `ConnectionBuilder` to set the header to your desired value. I don't believe the user agent presented during the authorization flow can be customized; I'm not aware of any mechanism provided by Chrome Custom Tabs to change this.

**Comment by chrismash2** [comment]
On the iOS library your custom agent class just gets provided the request URL that should start the flow and you let it know once the auth code has been received, so it avoids it getting booted out to another app/browser, which is the behaviour I'm after, but it looks like I might need to fork and add this myself?

**Comment by iainmcgin** [comment]
Ah, I misunderstood you. With AppAuth-Android, the closest equivalent is [control over browser selection](https://github.com/openid/AppAuth-Android/#controlling-which-browser-is-used-for-authorization). It sounds like you are unhappy with the use of custom tabs and instead want to use an embedded web view. We do not support the use of embedded web views, as this compromises the user's security during third party authorization.

**Comment by chrismash2** [comment]
It's not a web view I want to use, but to do it with almost no UI at all due to my specific use case. It's not really a true OAuth flow for this reason, but follows the same steps otherwise. I think I've found that it's possible to do what I want without changing anything as the request object that's constructed has the full URL that needs to be visited, so I'll close this issue now, thanks for your comments!

**Comment by hammad-tariq** [comment]
> It's not a web view I want to use, but to do it with almost no UI at all due to my specific use case. It's not really a true OAuth flow for this reason, but follows the same steps otherwise. I think I've found that it's possible to do what I want without changing anything as the request object that's constructed has the full URL that needs to be visited, so I'll close this issue now, thanks for your comments!

@chrismash2 Can you please provide some details ? We are using Keycloack as Identity provider and we don't want to use Custom tabs/Browser.

**Comment by chrismash2** [comment]
@hammad-tariq My setup was that I wanted to run the OIDC flow within the same app, with passwordless authentication (FIDO) to authenticate.

What I did was build up an authorization request with `AuthorizationRequest.Builder` and then use `toUri` to get the URI, which I could then use to make a network request without AppAuth's help. When I eventually got the 302 (which I avoided following the redirect on) I used `AuthorizationResponse.Builder` to create the response and then perform the token request with AppAuth.

That might not be exactly what you want to do for your use case but hopefully there's some tips in there.
